testkit/src/handler.rs

```diff
@@ -0,0 +1,433 @@
+        Status::Ok,
+        Header(ContentType::json()),
+    )))
+}
+
+trait TestKitHandler {
+    fn handle_status(&self, req: &mut Request) -> IronResult<Response>;
+    fn handle_create_block(&mut self, req: &mut Request) -> IronResult<Response>;
+    fn handle_rollback(&mut self, req: &mut Request) -> IronResult<Response>;
+}
+
+impl TestKitHandler for TestKit {
+    fn handle_status(&self, _: &mut Request) -> IronResult<Response> {
+        use exonum::helpers::Height;
+
+        #[derive(Debug, Serialize, Deserialize)]
+        struct TestKitStatus {
+            height: Height,
+            configuration: TestNetworkConfiguration,
+            next_configuration: Option<TestNetworkConfiguration>,
+        }
+
+        let status = TestKitStatus {
+            height: self.height(),
+            configuration: self.configuration_change_proposal(),
+            next_configuration: self.next_configuration().cloned(),
+        };
+        ok_response(&status)
+    }
+
+    fn handle_create_block(&mut self, req: &mut Request) -> IronResult<Response> {
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        struct CreateBlockRequest {
+            tx_hashes: Option<Vec<crypto::Hash>>,
+        }
+
+        let req = match req.get::<bodyparser::Struct<CreateBlockRequest>>() {
+            Ok(Some(req)) => req,
+            Ok(None) => CreateBlockRequest { tx_hashes: None },
+            Err(e) => Err(ApiError::BadRequest(e.to_string()))?,
+        };
+
+        let block_info = if let Some(tx_hashes) = req.tx_hashes {
+            let maybe_missing_tx = tx_hashes.iter().find(|h| !self.is_tx_in_pool(h));
+            if let Some(missing_tx) = maybe_missing_tx {
+                Err(ApiError::BadRequest(format!(
+                    "Transaction not in mempool: {}",
+                    missing_tx.to_string()
+                )))?;
+            }
+
+            // NB: checkpoints must correspond 1-to-1 to blocks.
+            self.checkpoint();
+            self.create_block_with_tx_hashes(&tx_hashes)
+        } else {
+            self.checkpoint();
+            self.create_block()
+        };
+
+        ok_response(&block_info)
+    }
+
+    fn handle_rollback(&mut self, req: &mut Request) -> IronResult<Response> {
+        let params = req.extensions.get::<Router>().unwrap();
+
+        let height: u64 = match params.find("height") {
+            Some(height_str) => height_str
+                .parse()
+                .map_err(|e: ParseIntError| ApiError::BadRequest(e.to_string()))?,
+            None => Err(ApiError::BadRequest(
+                "Required request parameter is missing: height".to_string(),
+            ))?,
+        };
+        if height == 0 {
+            Err(ApiError::BadRequest(
+                "Cannot rollback past genesis block".into(),
+            ))?;
+        }
+
+        if self.height().0 >= height {
+            let rollback_blocks = (self.height().0 - height + 1) as usize;
```

Comment:
It might be the matter of taste, but I think it's strange. We're using `Height` starting with `0`, but you have to pass `1` when you need to rollback to genesis block.
```rust
// Test roll-back to the genesis block
        request::delete(
            "http://localhost:3000/v1/blocks/1",
            Headers::new(),
            &handler,
        ).unwrap();
        {
            let testkit = testkit.read().unwrap();
            assert_eq!(testkit.height(), Height(0)); /// <- What? I want "1" here!
        }
```

Reply:
I'm not entirely sure that `DELETE` method in general is the best choice here (maybe, use something like `POST .../testkit/v1/rollback`?), but if we stick with it, it seems natural to me that `DELETE .../blocks/1` deletes the first block (i.e., the block immediately after the genesis block - the same URL scheme the explorer is using).

Reply:
Yep, it's okay when you read docs, but first time I was like wtf, it's mistake!